### ISSUE TITLE
Add a configuration file to determine the order in which migrations are run

### ISSUE THREAD

**Issue by ColinLaws** [laravel/ideas#1482]
I recently became frustrated with the process of refactoring the database design of an application I've been working on. During the process, I added, renamed, and modified relationships between multiple tables. The process for re-ordering these migrations essentially consists of having to rename the files over and over again. 

It would be cleaner and faster to have migration files drop the current timestamp prefix, and rely on a configuration file to determine the order in which migrations run. 

You could migrate an existing Laravel application to this structure by parsing the name of each migration file, and ordering the newly renamed files in the configuration based on the prefix that has been removed.

**Comment by ColinLaws** [comment]
I've been thinking a lot about what @JeffreyWay meant when he said, "reduce mental debt". I think it makes a lot of sense to just visually cut down on how much garbage there is visually when you open up the migrations folder. I know it's been like this for a while, and there are plenty of good reasons it should be this way, but I thought it was worth discussing how it could be improved.

**Comment by ColinLaws** [comment]
Right, there shouldn't really be vast changes in the structure of migrations files after release. Any subsequent migrations would make the necessary changes. Using a config to determine the order of migrations would really only be beneficial in cases like mine. I've been working on a Laravel based web app for over two years and recently had to make some pretty large changes to the migrations.

**Comment by sisve** [comment]
I agree that what migrations execute, and what you need in production, often differs. I often execute my DDL, and then mark the migrations as run manually, just to enforce in-place modifications. You do not want to accidentally rebuild a 100+ gb table...

However, I think that this idea would add to the mental debt; you would need to keep in mind how the project is configured and what configuration files says for every migration you write.

I also think you've misunderstood the migration structure. You're not supposed to edit old migration files or reorder them. Every migration is a step executed against the database to get it to its current structure. Every change in structure would be another migration file. You can of course build a migration file that does several changes at once, but you shouldn't modify the migration file once it's published (pushed to git) since it may have been executed by someone already.

If you modify old ones, how would another developer run the down() methods to revert changes? If you've changed the migration, then your new logic wouldn't restore the database structure to what it was when the developer ran up(), because the code may have changed between the up() and the down().

What you describe sounds like a way to configure the current database schema, a way to say "it should look like this", ignoring the history and how you ended up with the current state. There's currently no easy way to do this in Eloquent; there's no metadata available of what columns you expect to find in your database, so there's nothing to compare to the existing database to say "these tables with these columns are missing, and these columns needs to change data type".

We could maybe try and run all migrations to build a in-memory structure of a database, and use this to compare with the real database to figure out what changes are needed. I'm not sure this is possible or feasible with the way we currently write migrations, specially when migrations are just php files that could do anything and it would be unsafe to blindly execute them again. This is however a totally different idea.

TL;DR: I disagree with this idea.

**Comment by ColinLaws** [comment]
> I also think you've misunderstood the migration structure. You're not supposed to edit old migration files or reorder them. Every migration is a step executed against the database to get it to its current structure. Every change in structure would be another migration file. You can of course build a migration file that does several changes at once, but you shouldn't modify the migration file once it's published (pushed to git) since it may have been executed by someone already.

Yep, that's why I said it only makes sense before production. It would benefit me not having to rename files just to change the order in which they execute.

**Comment by sisve** [comment]
If you're in a state where you can freely reorder migrations without causing problems; couldn't you just have one large migration file that does everything, and move around code there?

**Comment by ColinLaws** [comment]
Yes, I could, but then I would have a massive migration file. Is that what people are doing for their initial database migration? To me, it makes sense to create a migration for every table being created, even if it's the initial migration.

**Comment by fitztrev** [comment]
> then I would have a massive migration file. Is that what people are doing for their initial database migration? To me, it makes sense to create a migration for every table being created, even if it's the initial migration.

Yes, that's how I do it. Each deploy to production has at most 1 new migration file that needs to run. All new database schema/data are consolidated in that.

**Comment by ColinLaws** [comment]
Gotcha, I guess that's not really a big deal having it in one file. I considered that approach when I began this project two years ago. Well, if there isn't anyone that agrees with me, I see no reason to continue fighting for it. It wouldn't really make sense to reorder migrations after they've been made.

**Comment by drbyte** [comment]
> then I would have a massive migration file. Is that what people are doing for their initial database migration? To me, it makes sense to create a migration for every table being created, even if it's the initial migration.

We use both approaches. Depends on the project.

**Comment by ColinLaws** [comment]
For the record, this is essentially my idea (don't mind the flag post comment not being perfect).

![migrations](https://user-images.githubusercontent.com/10246886/51193006-cf71ce00-18b5-11e9-8ed4-aa370704feca.PNG)
